rootfs/etc/nginx/lua/plugins.lua

```diff
@@ -0,0 +1,48 @@
+local string_format = string.format
+local new_tab = require "table.new"
+local ngx_log = ngx.log
+local INFO = ngx.INFO
+local ERR = ngx.ERR
+
+local _M = {}
+local MAX_NUMBER_OF_PLUGINS = 10000
+-- TODO: is this good for a dictionary?
+local plugins = new_tab(MAX_NUMBER_OF_PLUGINS, 0)
+
+local function load_plugin(name)
+  local path = string_format("plugins.%s.main", name)
+
+  local ok, plugin = pcall(require, path)
+  if not ok then
+    ngx_log(ERR, string_format("error loading plugin \"%s\": %s", path, plugin))
+    return
+  end
+
+  plugins[name] = plugin
```

Comment:
What about discovering plugin supported function on load, and avoid `plugin[phase] ` during request?

So we could only call required plugins.

Reply:
@wayt I think you are referring to what I describe in "later optimization" section in https://github.com/Shopify/ingress/pull/206? I wanna do it later, in the next iteration.